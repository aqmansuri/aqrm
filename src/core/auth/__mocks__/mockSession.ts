import { Session } from '../../models';

export const mockSession: Session = {
  token: '3884915llf950',
  user: {
    id: 42,
    firstName: 'Joe',
    lastName: 'Tester',
    email: 'test@test.org',
  },
};
