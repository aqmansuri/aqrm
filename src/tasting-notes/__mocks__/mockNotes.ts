import { TastingNote } from '../../shared/models';

export const mockNotes: TastingNote[] = [
  {
    id: 4,
    brand: 'Lipton',
    name: 'Yellow Label',
    notes: 'Overly acidic, highly tannic flavor',
    rating: 4,
    teaCategoryId: 1,
  },
  {
    id: 73,
    brand: 'Bently',
    name: 'Brown Label',
    notes: 'Essentially OK',
    rating: 3,
    teaCategoryId: 2,
  },
  {
    id: 42,
    brand: 'Lipton',
    name: 'Yellow Label',
    notes: 'Overly acidic, highly tannic flavor',
    rating: 1,
    teaCategoryId: 3,
  },
];
