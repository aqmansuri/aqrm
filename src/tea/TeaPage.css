.tea-grid ion-card {
  height: 100%;
}

.tea-grid ion-col {
  margin-bottom: 1em;
}
