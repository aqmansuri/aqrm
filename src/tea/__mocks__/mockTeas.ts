import { Tea } from '../../shared/models';

export const expectedTeas = [
  {
    id: 1,
    name: 'Green',
    image: require(`../../assets/images/green.jpg`).default,
    description: 'Green tea description.',
  },
  {
    id: 2,
    name: 'Black',
    image: require(`../../assets/images/black.jpg`).default,
    description: 'Black tea description.',
  },
  {
    id: 3,
    name: 'Herbal',
    image: require(`../../assets/images/herbal.jpg`).default,
    description: 'Herbal Infusion description.',
  },
  {
    id: 4,
    name: 'Oolong',
    image: require(`../../assets/images/oolong.jpg`).default,
    description: 'Oolong tea description.',
  },
  {
    id: 5,
    name: 'Dark',
    image: require(`../../assets/images/dark.jpg`).default,
    description: 'Dark tea description.',
  },
  {
    id: 6,
    name: 'Puer',
    image: require(`../../assets/images/puer.jpg`).default,
    description: 'Puer tea description.',
  },
  {
    id: 7,
    name: 'White',
    image: require(`../../assets/images/white.jpg`).default,
    description: 'White tea description.',
  },
  {
    id: 8,
    name: 'Yellow',
    image: require(`../../assets/images/yellow.jpg`).default,
    description: 'Yellow tea description.',
  },
];

export const resultTeas = () => {
  return expectedTeas.map((t: Tea) => {
    const tea = { ...t };
    // @ts-ignore
    delete tea.image;
    return tea;
  });
};
