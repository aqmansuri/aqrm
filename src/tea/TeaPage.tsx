import { useEffect, useState } from 'react';
import {
  IonCard,
  IonCardContent,
  IonCardHeader,
  IonCardTitle,
  IonCol,
  IonContent,
  IonGrid,
  IonHeader,
  IonImg,
  IonPage,
  IonRow,
  IonTitle,
  IonToolbar,
  IonButtons,
  IonButton,
  IonIcon,
} from '@ionic/react';
import { logOutOutline } from 'ionicons/icons';
import { useHistory } from 'react-router';
import { useAuthentication } from '../core/auth';
import { Tea } from '../shared/models';
import { useTea } from './useTea';
import './TeaPage.css';

export const listToMatrix = (teaArray: Tea[]): Tea[][] => {
  let teaMatrix: Tea[][] = [];

  let row: Tea[] = [];
  teaArray.forEach(tea => {
    row.push(tea);
    if (row.length === 4) {
      teaMatrix.push(row);
      row = [];
    }
  });

  if (row.length) teaMatrix.push(row);

  return teaMatrix;
};

const TeaPage: React.FC = () => {
  const [teas, setTeas] = useState<Tea[]>([]);
  const { logout } = useAuthentication();
  const history = useHistory();
  const { getTeas } = useTea();

  useEffect(() => {
    (async () => {
      const teas = await getTeas();
      setTeas(teas);
    })();
  }, [getTeas]);

  const handleLogout = async () => {
    await logout();
    history.replace('/login');
  };

  return (
    <IonPage>
      <IonHeader>
        <IonToolbar>
          <IonTitle>Tea</IonTitle>
          <IonButtons slot="end">
            <IonButton onClick={() => handleLogout()}>
              <IonIcon slot="icon-only" icon={logOutOutline} />
            </IonButton>
          </IonButtons>
        </IonToolbar>
      </IonHeader>
      <IonContent fullscreen>
        <IonHeader collapse="condense">
          <IonToolbar>
            <IonTitle size="large">Tea</IonTitle>
          </IonToolbar>
        </IonHeader>

        <IonGrid className="tea-grid">
          {listToMatrix(teas).map((row, idx) => (
            <IonRow
              key={idx}
              className="ion-justify-content-center ion-align-items-stretch"
            >
              {row.map(tea => (
                <IonCol size="12" sizeMd="6" sizeXl="3" key={tea.id}>
                  <IonCard>
                    <IonImg src={tea.image} />
                    <IonCardHeader>
                      <IonCardTitle>{tea.name}</IonCardTitle>
                    </IonCardHeader>
                    <IonCardContent>{tea.description}</IonCardContent>
                  </IonCard>
                </IonCol>
              ))}
            </IonRow>
          ))}
        </IonGrid>
      </IonContent>
    </IonPage>
  );
};

export default TeaPage;
