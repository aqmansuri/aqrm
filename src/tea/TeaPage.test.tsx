import { render, waitFor } from '@testing-library/react';
import TeaPage, { listToMatrix } from './TeaPage';
import { expectedTeas } from './__mocks__/mockTeas';

const mockTeas = expectedTeas;
jest.mock('./useTea', () => ({
  useTea: () => ({
    getTeas: jest.fn(() => Promise.resolve(mockTeas)),
    getTeaById: jest.fn(),
  }),
}));

describe('<TeaPage />', () => {
  it('displays the header', async () => {
    const { container } = render(<TeaPage />);
    await waitFor(() => expect(container).toHaveTextContent(/Tea/));
  });

  it('renders consistently', async () => {
    const { asFragment } = render(<TeaPage />);
    await waitFor(() => expect(asFragment()).toMatchSnapshot());
  });

  describe('initialization', () => {
    it('makes a tea matrix', () => {
      const teaMatrix = [
        [expectedTeas[0], expectedTeas[1], expectedTeas[2], expectedTeas[3]],
        [expectedTeas[4], expectedTeas[5], expectedTeas[6], expectedTeas[7]],
      ];
      expect(listToMatrix(expectedTeas)).toEqual(teaMatrix);
    });
  });
});
