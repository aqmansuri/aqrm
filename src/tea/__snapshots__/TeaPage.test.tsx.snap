// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TeaPage /> renders consistently 1`] = `
<DocumentFragment>
  <div
    class="ion-page"
  >
    <ion-header>
      <ion-toolbar>
        <ion-title>
          Tea
        </ion-title>
        <ion-buttons
          slot="end"
        >
          <ion-button>
            <ion-icon
              icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><title>Log Out</title><path d='M304 336v40a40 40 0 01-40 40H104a40 40 0 01-40-40V136a40 40 0 0140-40h152c22.09 0 48 17.91 48 40v40M368 336l80-80-80-80M176 256h256' stroke-linecap='round' stroke-linejoin='round' class='ionicon-fill-none ionicon-stroke-width'/></svg>"
              slot="icon-only"
            />
          </ion-button>
        </ion-buttons>
      </ion-toolbar>
    </ion-header>
    <ion-content
      fullscreen="true"
    >
      <ion-header
        collapse="condense"
      >
        <ion-toolbar>
          <ion-title
            size="large"
          >
            Tea
          </ion-title>
        </ion-toolbar>
      </ion-header>
      <ion-grid
        class="tea-grid"
      />
    </ion-content>
  </div>
</DocumentFragment>
`;
